### ISSUE TITLE
Failed to resolve: com.github.AppIntro:AppIntro:6.1.0

### ISSUE THREAD

**Issue by Ojd2000** [AppIntro/AppIntro#957]
## 🐛 Describe the bug

I added jitpack to my repositories and AppIntro to my dependencies, but still gradle fails to resolve it.
I'm not using the deprecated jcenter as a repository, but even if I add it, it fails. My repositories are all added in the root `build.gradle` file, since trying to add any to the module's file generates a `GradleScriptException` with the following error:

> Build was configured to prefer settings repositories over project repositories but repository 'maven' was added by build file 'app/build.gradle'

I don't really know why this is the case, so I put here the error.

## ⚠️ Current behavior
Gradle fails to resolve `com.github.AppIntro:AppIntro:6.1.0`.
Build succeeds.

## ✅ Expected behavior
Gradle succeeding to resolve the dependency.

## 💣 Steps to reproduce

1. Create a new Android Studio project
2. Add `maven { url 'https://jitpack.io' }` repository inside `buildscript` in root's `build.gradle` file
3. Add `'com.github.AppIntro:AppIntro:6.1.0'` to the dependencies in module's `build.gradle` file
4. Sync repositories

## 📑 Your Code

 My `build.gradle` root file:
```
buildscript {
    repositories {
        google()
        mavenCentral()
        maven { url 'https://jitpack.io' }
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:7.0.0'
        classpath 'org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.21'
    }
}
task clean(type: Delete) {
    delete rootProject.buildDir
}
```

My `build.gradle` module file:
```
plugins {
    id 'com.android.application'
    id 'kotlin-android'
}
android {
    compileSdk 31

    defaultConfig {
        applicationId 'myappid'
        minSdk 28
        targetSdk 31
        versionCode 1
        versionName '1.0'

        testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
    buildFeatures {
        viewBinding true
    }
}
dependencies {
    implementation 'androidx.core:core-ktx:1.6.0'
    implementation 'androidx.appcompat:appcompat:1.3.1'
    implementation 'androidx.recyclerview:recyclerview:1.2.1'
    implementation 'com.google.android.material:material:1.4.0'
    implementation 'androidx.constraintlayout:constraintlayout:2.1.0'
    implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.3.1'
    implementation 'androidx.lifecycle:lifecycle-viewmodel-ktx:2.3.1'
    implementation 'androidx.navigation:navigation-fragment-ktx:2.3.5'
    implementation 'androidx.navigation:navigation-ui-ktx:2.3.5'
    implementation 'androidx.preference:preference-ktx:1.1.1'
    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
    testImplementation 'junit:junit:4.13.2'
    androidTestImplementation 'androidx.test.ext:junit:1.1.3'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'

    implementation 'com.github.AppIntro:AppIntro:6.1.0'
}
```

## 📱 Tech info
 - AppIntro Version: 6.1.0
 - Android Studio Version: Arctic Fox 2020.3.1.22
 - Gradle Version: 7.1.1
 - Development Machine OS: Arch Linux (kernel 5.13.7.zen1)

**Comment by cortinico** [comment]
The problem is that you need to move the Jitpack repository you declared.
Right now you have inside the `buildscript{}` block. You should remove it from there and add it to the `repositories{}` block that you find at the top level (either in your top level build.gradle or inside the module one).

For example, if you add the following to your module `build.gradle`:
```
repositories {
    google()
    mavenCentral()
    maven { url 'https://jitpack.io' }
}
```
It will work correctly

**Comment by Ojd2000** [comment]
@cortinico Actually, if I do what you say and add the `repositories{}` block to either of the two `build.gradle`, it just gives me the error I mentioned in my [previous comment](https://github.com/AppIntro/AppIntro/issues/957#issue-958899444).

To avoid that error, I found out I just need to change `repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)` to `repositoriesMode.set(RepositoriesMode.PREFER_SETTINGS)` in my `settings.gradle`, and it does replace it with a simple warning.
The problem is that gradle is still not able to resolve the AppIntro dependency, so this makes me think thatthe place I put the Jitpack repository in is irrelevant to my problem.

**Comment by cortinico** [comment]
What you're describing is weird. Are you able to provide a reproducer of some form?

> The problem is that gradle is still not able to resolve the AppIntro dependency, so this makes me think that the place I put the Jitpack repository in is irrelevant to my problem.

Definitely is not. If you place Jitpack inside the `buildscript` block, that repository will be used only for build dependencies (aka Gradle plugins or `classpath` dependencies). If you place it in the top level `repositories` block, that repository will be used for runtime/compile dependencies (like AppIntro).

> I found out I just need to change repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS) to repositoriesMode.set(RepositoriesMode.PREFER_SETTINGS) in my settings.gradle, and it does replace it with a simple warning.

Ideally you should not be doing this. Like this is not addressing the root cause of your problem.

**Comment by Ojd2000** [comment]
@cortinico I figured out how to solve my issue... I added the repository in `settings.gradle` instead of `build.gradle`. This is my `settings.gradle` file:
```
dependencyResolutionManagement {
    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
    repositories {
        google()
        mavenCentral()
        maven { url 'https://jitpack.io' }
    }
}
rootProject.name = "My Application"
include ':app'
```

> Ideally you should not be doing this. Like this is not addressing the root cause of your problem.

Indeed I should not, I was just saying that forcing gradle to accept repositories outside the `buildscript` block where I tried was not working.

> Are you able to provide a reproducer of some form?

So I re-created another Android Studio project in version Arctic Fox 2020.3.1.22 just to confirm that this is universal (at least in this version). These are my steps:

1. Create a new project with an empty activity, just to add minimum code.
2. Add `maven { url 'https://jitpack.io' }` repository inside, outside `buildscript` in root's `build.gradle` or in module's in turns.
3. Add `'com.github.AppIntro:AppIntro:6.1.0'` to the `dependencies` in module's `build.gradle` file.
4. Sync repositories.
5. This should bring to the same issues I mentioned in my other comments.

Based on comments I briefly read in these days in the Internet, I think that this is a new way Android Studio (or Gradle) handles projects, but I'm not much experienced with Android Studio and Gradle, so I may be completely wrong.

**Comment by cortinico** [comment]
> These are my steps:

I've just tested them once more. It works fine for me.
For your reference those are the Gradle files (I've cleaned them up a bit):

Top level `build.gradle`.
```groovy
buildscript {
    ext.kotlin_version = "1.5.0"
    repositories {
        google()
        mavenCentral()
    }
    dependencies {
        classpath "com.android.tools.build:gradle:4.2.1"
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
    }
}

allprojects {
    repositories {
        google()
        mavenCentral()
        maven { url 'https://jitpack.io' }
    }
}
```

App level `build.gradle`
```groovy
plugins {
    id 'com.android.application'
    id 'kotlin-android'
}

android {
    compileSdkVersion 30
    defaultConfig {
        applicationId "com.example.myapplication"
        minSdkVersion 16
        targetSdkVersion 30
        versionCode 1
        versionName "1.0"
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
}

dependencies {
    implementation 'com.github.AppIntro:AppIntro:6.1.0'
    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
    implementation 'androidx.core:core-ktx:1.6.0'
    implementation 'androidx.appcompat:appcompat:1.3.1'
    implementation 'com.google.android.material:material:1.3.0'
    implementation 'androidx.constraintlayout:constraintlayout:2.0.4'
}
```

**Comment by Ojd2000** [comment]
@cortinico I think I figured it out.
I tried compiling the project with a bunch of different gradle versions: `4.2.1`, `4.2.2`, and `7.1.1`; they all behaved the same.

The main problem is the `dependencyResolutionManagement` block in `settings.gradle`: commenting it makes gradle compile, if you add repos where you were saying. If that block is present and the repository mode is set to prefer settings repositories over project ones, that means that all repositories (for compile and runtime dependencies) must be in the `settings.gradle` file as I showed in my previous comment.

I also noticed you are using gradle `4.2.1` in the file you provided, while I am using `7.0.0` (in my first comment) and then `7.1.1` for some other tests, so I suppose that is enough for Android Studio/Gradle not to create the `dependencyResolutionManagement` block (or using a different repository mode) at project creation.
